### PR TITLE
Added Jupyter Chapter 2.2 (TapTweak)

### PR DESCRIPTION
**Pull request by jachiang**
Note: [Previous feedback](https://gist.github.com/jachiang/d6854c13647199973560440e07e2e3c1) on this chapter.

### CODE REVIEW

**Comment by elichai** [comment]
btw, when I run this I get:
```
AssertionError                            Traceback (most recent call last)
<ipython-input-1-579dc5073092> in <module>
      1 # Add Path to Bitcoin Core Testframework Library.
----> 2 import util
      3 
      4 from test_framework.key import ECKey, ECPubKey, SECP256K1_ORDER
      5 from test_framework.key import generate_schnorr_nonce, aggregate_schnorr_nonces

~/gits/taproot-workshop/util.py in <module>
     11 SOURCE_DIRECTORY = config["path"]["SOURCE_DIRECTORY"]
     12 
---> 13 assert not SOURCE_DIRECTORY == '', 'SOURCE_DIRECTORY not configured! Edit config.ini to configure SOURCE_DIRECTORY.'
     14 
     15 print("Source directory configured as {}".format(SOURCE_DIRECTORY))

AssertionError: SOURCE_DIRECTORY not configured! Edit config.ini to configure SOURCE_DIRECTORY.

```

File: Solutions/2.2-TapTweak-Solutions.ipynb [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+    "k0 = generate_schnorr_nonce()\n",
+    "k1 = generate_schnorr_nonce()\n",
+    "R0 = k0.get_pubkey()\n",
+    "R1 = k1.get_pubkey()\n",
+    "R_agg, negated = aggregate_schnorr_nonces([R0, R1])\n",
+    "\n",
+    "# C2) Signing and signature aggregation.\n",
+    "msg = hashlib.sha256(b'msg').digest()\n",
+    "\n",
+    "# One person must tweak keys.\n",
```
Comment (elichai) [inline Solutions/2.2-TapTweak-Solutions.ipynb]: Technically you could even tweak the signature.
but yeah, if one of the participants tweak their keys that also works.
(don't know how to libsecp musig implementation will handle this)

File: 2.2-TapTweak.ipynb [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+    "* Part 3 - TapTweak Script Commitments\n",
+    "\n",
+    "\n",
+    "## Part 1: Tweaking the Public Key\n",
+    "\n",
+    "The script commitment scheme of taproot is based on tweaking the public key. \n",
+    "\n",
+    "* `[01] [33B Tweaked Public Key]`\n",
+    "\n",
+    "Tweaking a public key means to alter it with a value (the tweak) so that it remains spendable with knowledge of the original private key and tweak. \n",
```
Comment (elichai) [inline 2.2-TapTweak.ipynb]:
```suggestion
    "Tweaking a public key means to alter it with a value (the tweak) such that it remains spendable with knowledge of the original private key and tweak. \n",
```

File: 2.2-TapTweak.ipynb [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Example 2.1- Modifying the tweak for a tweaked public key Q.\n",
+    "\n",
+    "* A) Tweaking the public key to obtain: `Point Q`\n",
+    "* B) Create a new tweak for the same point Q: `t'`\n",
+    "* C) Solve for `x'` so that `x'G + t'G = Q`"
```
Comment (elichai) [inline 2.2-TapTweak.ipynb]:
```suggestion
    "* C) Solve for `x'` such that `x'G + t'G = Q`"
```

File: 2.2-TapTweak.ipynb (hunk 2 above)
Comment (elichai) [reply 2.2-TapTweak.ipynb]: nit

File: 2.2-TapTweak.ipynb (hunk 3 above)
Comment (elichai) [reply 2.2-TapTweak.ipynb]: nit

**Comment by jnewbery** [comment]
@elichai 
> btw, when I run this I get:
...
`AssertionError: SOURCE_DIRECTORY not configured! Edit config.ini to configure SOURCE_DIRECTORY.`

Can you run through the `0.1-test-notebook.ipynb` notebook and verify that setting config.ini works for you. Also note the developer hint here: https://github.com/bitcoinops/taproot-workshop/pull/12#issue-313357263 so you don't need to change config.ini more than once.

**Comment by jachiang** [comment]
Updated for [this](https://github.com/ChaincodeResidency/bitcoin/pull/24) bitcoin taproot branch - includes updated MuSig python module.

**Comment by jnewbery** [comment]
rebased on master

**Comment by jnewbery** [comment]
rebased on master

File: Solutions/2.2-TapTweak-Solutions.ipynb (hunk 1 above)
Comment (jnewbery) [reply Solutions/2.2-TapTweak-Solutions.ipynb]: Let's leave this as one person tweaking their key when signing

File: 2.2-TapTweak.ipynb (hunk 2 above)
Comment (jnewbery) [reply 2.2-TapTweak.ipynb]: fixed

File: 2.2-TapTweak.ipynb (hunk 3 above)
Comment (jnewbery) [reply 2.2-TapTweak.ipynb]: fixed

**Comment by jnewbery** [comment]
I've run through this and all the code runs successfully, so I'm going to merge now.

I've addressed all of elichai's comments. I still plan to review this thoroughly myself.

**Pull request merged**